tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+#  (PART) Conclusion {-}
+
+# Tradeoffs
+
+::: {.alert .alert-success} 
+This somewhat opinionated discussion was added by [Max Held](https://www.maxheld.de).
+Your mileage may vary.
+:::
+
+Test doubles such as mocks or fakes can help you isolate your tests from the state of external HTTP API.
+This can be a good thing, because it allows you to test your code against a rare responses, or when the API is not available.
+
+Unfortunately, mocking and faking add their own complexity to a project.
+Respectively, you have to carry expected responses or even (rudimentary) API behavior in your code base.
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
```

Comment:
Isn't this a repetition of https://books.ropensci.org/http-testing/graceful.html#graceful-code?

Could you add some mock tests (ahah)? I.e. two functions calling an API, a version without and a version with utility functions?

Reply:
> Isn't this a repetition of https://books.ropensci.org/http-testing/graceful.html#graceful-code?

Yes, sorry -- added a reference.
I think the angle here is somewhat different, drawing the parallel to side effects, which we generally want to avoid and minimise/concentrate in as few functions as possible.

Reply:
> Could you add some mock tests (ahah)? I.e. two functions calling an API, a version without and a version with utility functions?

I'm not sure I fully understand.
I think the hypothetical below `get_doi_url()` might be this example, where I mention that it (and other upstream functions) won't need to query the API themselves, but can rely on `get_doi_handle()` instead.
I guess this is the kind of thing that would benefit from integrating the examples with all of the text as you mentioned.

Reply:
oops I meant mock code not mock tests!!